system/system.go

```diff
@@ -80,128 +80,119 @@ func New(c *cli.Context) *System {
 		NewProcess:  NewDefProcess,
 		NewGossfile: NewDefGossfile,
 	}
-	// FIXME: Detect-os needs to be refactored in a consistent way
-	// Also, cache should be its own object
 	sys.detectService()
+	sys.detectPackage(c)
+	return sys
+}
 
+// detectPackage adds the correct package creation function to a System struct
+func (sys *System) detectPackage(c *cli.Context) {
 	p := c.GlobalString("package")
+	if p != "deb" && p != "apk" && p != "pacman" && p != "rpm" {
+		p = DetectPackageManager()
+	}
 	switch p {
-	case "rpm":
-		sys.NewPackage = NewRpmPackage
 	case "deb":
 		sys.NewPackage = NewDebPackage
-	case "alpine":
+	case "apk":
 		sys.NewPackage = NewAlpinePackage
 	case "pacman":
 		sys.NewPackage = NewPacmanPackage
 	default:
-		sys.NewPackage = detectPackage()
-	}
-
-	return sys
-}
-
-func detectPackage() func(string, *System, util2.Config) Package {
-	switch {
-	case isRpm():
-		return NewRpmPackage
-	case isDeb():
-		return NewDebPackage
-	case isAlpine():
-		return NewAlpinePackage
-	case isPacman():
-		return NewPacmanPackage
-	default:
-		return NewNullPackage
+		sys.NewPackage = NewRpmPackage
 	}
 }
 
-func (s *System) detectService() {
-	switch {
-	case util.IsRunningSystemd():
-		s.NewService = NewServiceDbus
-	case isUbuntu():
-		s.NewService = NewServiceUpstart
-	case isAlpine():
-		s.NewService = NewAlpineServiceInit
-	case isArch():
-		s.NewService = NewServiceDbus
+// detectService adds the correct service creation function to a System struct
+func (sys *System) detectService() {
+	switch DetectService() {
+	case "upstart":
+		sys.NewService = NewServiceUpstart
+	case "systemd":
+		sys.NewService = NewServiceDbus
+	case "alpineinit":
+		sys.NewService = NewServiceAlpineInit
 	default:
-		s.NewService = NewServiceInit
+		sys.NewService = NewServiceInit
 	}
 }
 
-func isUbuntu() bool {
-	if b, err := ioutil.ReadFile("/etc/lsb-release"); err == nil {
-		if bytes.Contains(b, []byte("Ubuntu")) {
-			return true
+// DetectPackageManager attempts to detect whether or not the system is using
+// "deb", "rpm", "apk", or "pacman" package managers. If none are found, it
+// attempts to fall back to detecting the running distro. If that fails, it
+// returns the empty string.
+func DetectPackageManager() string {
+	for _, manager := range []string{"deb", "rpm", "apk", "pacman"} {
+		if HasCommand(manager) {
+			return manager
```

Comment:
@aelsabbahy I'll reverse the order. I think we could have better performance gains by focusing elsewhere :) this seems so minor that it would be impossible to tell the difference.